src/container/srv_csum_recalc.c

```diff
@@ -0,0 +1,311 @@
+/**
+ * (C) Copyright 2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+#define D_LOGFAC	DD_FAC(csum)
+
+#include <daos/common.h>
+#include <daos/checksum.h>
+#include <daos_srv/daos_server.h>
+#include <daos_srv/container.h>
+#include <daos_srv/evtree.h>
+
+/*
+ * Checksum recalculation code. Called from vos_aggregate.c.
+ *
+ * Recalculation is driven by an array of csum_recalc structs,
+ * one per input segment. These segments are coalecsed into a single
+ * output segment by the overall aggregation process.
+ *
+ * The input data is held in a buffer that is specifed by a bio sg_list.
+ * The data for the output segment is place within the initial range of
+ * the buffer. Following this range, the additional data required for
+ * checksum data is stored. These additional segments, either prefix
+ * or suffix ranges of each input physical extent, are appended in order
+ * at the end of the buffer, with corresponding * entries in the bgsl's
+ * biov array.
```

Comment:
What's the purpose of these appended biovs? The original biovs already have all required information. (request data, prefix & postfix data from checksum calculation)

Reply:
I append them so I can store the actual write data separately at the first of the buffer.  This range is written using bio_write.  There is a a bio_writev, but it is not used anywhere and it does separate writes to media for each biov entry.  Best to keep it as a single contiguous range.

Reply:
I see, thanks for the explanation!